File: src/vip/synthesizer.py
"""AI synthesis with multiple backend support.

Supported backends:
- claude-cli: Local Claude Code CLI (default, no API key needed)
- anthropic: Claude API via anthropic SDK (needs ANTHROPIC_API_KEY)
- copilot-cli: GitHub Copilot CLI
"""

from __future__ import annotations

import os
import subprocess
from datetime import date

from vip.config import check_tool, load_config
from vip.templates import PROFILE_SYSTEM_PROMPT, CHANGE_DETECTION_PROMPT


def _get_backend() -> str:
    """Determine which AI backend to use.

    Priority:
    1. VIP_AI_BACKEND env var (explicit override)
    2. Config file setting
    3. Auto-detect: anthropic SDK > claude CLI > copilot CLI
    """
    # Explicit override
    backend = os.environ.get("VIP_AI_BACKEND", "").lower()
    if backend:
        return backend

    # Config file
    config = load_config()
    backend = config.get("ai_backend", "").lower()
    if backend:
        return backend

    # Auto-detect
    if os.environ.get("ANTHROPIC_API_KEY"):
        return "anthropic"
    if check_tool("claude"):
        return "claude-cli"
    if check_tool("gh") and _copilot_available():
        return "copilot-cli"

    raise RuntimeError(
        "No AI backend available. Options:\n"
        "  1. Install Claude Code CLI\n"
        "  2. Set ANTHROPIC_API_KEY env var (pip install anthropic)\n"
        "  3. Install GitHub Copilot CLI (gh copilot)"
    )


def _copilot_available() -> bool:
    """Check if GitHub Copilot CLI is available."""
    try:
        result = subprocess.run(
            ["gh", "copilot", "--help"],
            capture_output=True, text=True, timeout=5,
        )
        return result.returncode == 0
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return False


def _call_claude_cli(prompt: str, timeout: int = 120) -> str:
    """Call local Claude Code CLI."""
    result = subprocess.run(
        ["claude", "--print", "-p", prompt],
        capture_output=True, text=True, timeout=timeout,
    )
    if result.returncode != 0:
        raise RuntimeError(f"Claude CLI error: {result.stderr.strip()}")
    return result.stdout.strip()


def _call_anthropic_api(prompt: str) -> str:
    """Call Claude API directly via anthropic SDK."""
    try:
        import anthropic
    except ImportError:
        raise RuntimeError(
            "anthropic package not installed. Run: pip install anthropic"
        )

    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        raise RuntimeError("ANTHROPIC_API_KEY environment variable not set.")

    client = anthropic.Anthropic(api_key=api_key)
    config = load_config()
    model = config.get("anthropic_model", "claude-sonnet-4-20250514")

    message = client.messages.create(
        model=model,
        max_tokens=4096,
        messages=[{"role": "user", "content": prompt}],
    )
    return message.content[0].text.strip()


def _call_copilot_cli(prompt: str, timeout: int = 120) -> str:
    """Call GitHub Copilot CLI."""
    result = subprocess.run(
        ["gh", "copilot", "suggest", "-t", "shell", prompt],
        capture_output=True, text=True, timeout=timeout,
    )
    if result.returncode != 0:
        raise RuntimeError(f"Copilot CLI error: {result.stderr.strip()}")
    return result.stdout.strip()


def _call_backend(prompt: str, backend: str) -> str:
    """Route to the appropriate AI backend."""
    if backend == "claude-cli":
        return _call_claude_cli(prompt)
    elif backend == "anthropic":
        return _call_anthropic_api(prompt)
    elif backend == "copilot-cli":
        return _call_copilot_cli(prompt)
    else:
        raise RuntimeError(f"Unknown AI backend: {backend}")


def get_backend_name() -> str:
    """Get the name of the current AI backend (for display)."""
    try:
        return _get_backend()
    except RuntimeError:
        return "none"


def synthesize_profile(raw_data: str, sources: list[str] | None = None) -> str:
    """Generate a structured Markdown profile from raw data."""
    backend = _get_backend()
    prompt = f"{PROFILE_SYSTEM_PROMPT}\n\nRaw data:\n{raw_data}"

    profile = _call_backend(prompt, backend)

    today = date.today().isoformat()
    footer = f"\n\n---\n*Last updated: {today}*"
    if sources:
        source_list = ", ".join(sources[:10])
        footer += f"\n*Sources: {source_list}*"

    return profile + footer


def detect_changes(old_profile: str, new_data: str) -> str | None:
    """Compare old profile with new data to detect significant changes."""
    try:
        backend = _get_backend()
    except RuntimeError:
        return None

    prompt = CHANGE_DETECTION_PROMPT.format(
        old_profile=old_profile,
        new_data=new_data,
    )

    try:
        output = _call_backend(prompt, backend)
        if "NO_SIGNIFICANT_CHANGES" in output:
            return None
        return output
    except (subprocess.TimeoutExpired, RuntimeError):
        return None

File: lib/synthesizer.js
import { execFileSync, spawnSync } from 'child_process';
import fs from 'fs';
import os from 'os';
import path from 'path';
import { checkTool, loadConfig } from './config.js';
import { PROFILE_SYSTEM_PROMPT, CHANGE_DETECTION_PROMPT } from './templates.js';

function getBackend() {
  const envBackend = process.env.VIP_AI_BACKEND?.toLowerCase();
  if (envBackend) return envBackend;

  const config = loadConfig();
  if (config.ai_backend) return config.ai_backend.toLowerCase();

  if (checkTool('claude')) return 'claude-cli';
  if (process.env.ANTHROPIC_API_KEY || config.anthropic_api_key) return 'anthropic';
  if (checkTool('gh') && copilotAvailable()) return 'copilot-cli';

  throw new Error(
    'No AI backend available. Options:\n' +
    '  1. Install Claude Code CLI\n' +
    '  2. Set ANTHROPIC_API_KEY env var\n' +
    '  3. Install GitHub Copilot CLI (gh copilot)'
  );
}

function copilotAvailable() {
  try {
    execFileSync('gh', ['copilot', '--help'], { stdio: 'ignore', timeout: 5000 });
    return true;
  } catch {
    return false;
  }
}

function callClaudeCli(prompt, timeout = 300000) {
  // Write prompt to temp file, pass via -p flag reading from file
  const tmpFile = path.join(os.tmpdir(), `vip-prompt-${Date.now()}.txt`);
  try {
    fs.writeFileSync(tmpFile, prompt);
    // Try stdin pipe first
    const result = spawnSync('claude', ['--print'], {
      input: prompt,
      encoding: 'utf-8',
      timeout,
      stdio: ['pipe', 'pipe', 'pipe'],
      maxBuffer: 1024 * 1024 * 10,
    });
    if (result.error) {
      throw new Error(`Claude CLI error: ${result.error.message}`);
    }
    if (result.status !== 0) {
      const stderr = result.stderr?.trim() || '';
      const stdout = result.stdout?.trim() || '';
      // If stdin pipe failed, try with temp file
      const result2 = spawnSync('sh', ['-c', `cat "${tmpFile}" | claude --print`], {
        encoding: 'utf-8',
        timeout,
        stdio: ['pipe', 'pipe', 'pipe'],
        maxBuffer: 1024 * 1024 * 10,
      });
      if (result2.status === 0 && result2.stdout?.trim()) {
        return result2.stdout.trim();
      }
      throw new Error(`Claude CLI failed (exit ${result.status}): ${stderr || stdout || 'no output'}`);
    }
    return result.stdout.trim();
  } finally {
    try { fs.unlinkSync(tmpFile); } catch {}
  }
}

async function callAnthropicApi(prompt) {
  let anthropic;
  try {
    anthropic = await import('@anthropic-ai/sdk');
  } catch {
    throw new Error('Anthropic SDK not installed. Run: npm install @anthropic-ai/sdk');
  }

  const apiKey = process.env.ANTHROPIC_API_KEY || loadConfig().anthropic_api_key;
  if (!apiKey) throw new Error('ANTHROPIC_API_KEY not set. Run "vip init" or set the environment variable.');

  const config = loadConfig();
  const model = config.anthropic_model || 'claude-sonnet-4-20250514';

  const client = new anthropic.default({ apiKey });
  const message = await client.messages.create({
    model,
    max_tokens: 8192,
    messages: [{ role: 'user', content: prompt }],
  });

  return message.content[0].text.trim();
}

function callCopilotCli(prompt, timeout = 120000) {
  const result = execFileSync('gh', ['copilot', 'suggest', '-t', 'shell', prompt], {
    encoding: 'utf-8',
    timeout,
    stdio: ['pipe', 'pipe', 'pipe'],
  });
  return result.trim();
}

async function callBackend(prompt, backend) {
  if (backend === 'claude-cli') return callClaudeCli(prompt);
  if (backend === 'anthropic') return await callAnthropicApi(prompt);
  if (backend === 'copilot-cli') return callCopilotCli(prompt);
  throw new Error(`Unknown AI backend: ${backend}`);
}

export function getBackendName() {
  try {
    return getBackend();
  } catch {
    return 'none';
  }
}

export async function synthesizeProfile(rawData, sources) {
  const backend = getBackend();
  const prompt = `${PROFILE_SYSTEM_PROMPT}\n\nRaw data:\n${rawData}`;

  const profile = await callBackend(prompt, backend);

  const today = new Date().toISOString().slice(0, 10);
  let footer = `\n\n---\n*Last updated: ${today}*`;
  if (sources?.length) {
    footer += `\n*Sources: ${sources.slice(0, 10).join(', ')}*`;
  }

  return profile + footer;
}

export async function detectChanges(oldProfile, newData) {
  let backend;
  try {
    backend = getBackend();
  } catch {
    return null;
  }

  const prompt = CHANGE_DETECTION_PROMPT
    .replace('{old_profile}', oldProfile)
    .replace('{new_data}', newData);

  try {
    const output = await callBackend(prompt, backend);
    return output.includes('NO_SIGNIFICANT_CHANGES') ? null : output;
  } catch {
    return null;
  }
}
